### PR TITLE
HTTP Client implementation

### LINKED ISSUE #59: HTTP Client
ℹ️ This issue is part of Epic #57 

## Context

This module is responsible to communicate the [Pact API](https://api.chainweb.com/openapi/pact.html) with the Chainweb modules

## Objectives

- Sends the request body to [PACT](https://api.chainweb.com/openapi/pact.html) and returns the response.
- A contract will be defined and a default implementation will be made with :hackney (see the StellarSDK).

### PR DESCRIPTION
**Pull request by boterop**
## Description

Fixes #59 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## How it works
image

### CODE REVIEW

File: lib/chainweb/client/default.ex [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  @type headers :: [{binary(), binary()}, ...]
+  @type body :: binary()
+  @type success_response :: {:ok, status(), headers(), body()}
+  @type error_response :: {:error, status(), headers(), body()} | {:error, any()}
+  @type client_response :: success_response() | error_response()
+  @type parsed_response :: {:ok, map()} | {:error, Keyword.t()}
+
+  @impl true
+  def request(method, path, chain_id \\ "0", headers \\ [], body \\ "", opts \\ []) do
+    base_url = Network.base_url() |> String.replace("{chain_id}", chain_id)
```
Comment (juanhurtado10) [inline lib/chainweb/client/default.ex]: @santiagoboterokommit I'm not sure about the logic behind this code block. Why not just pass the `chain_id` as a parameter?

File: lib/chainweb/client.ex [hunk 2]
```diff
@@ -0,0 +1,17 @@
+defmodule Kadena.Chainweb.Client do
+  @moduledoc """
+  Specifies the API for processing HTTP requests in the Kadena network.
+  """
+  alias Kadena.Chainweb.Client
+
+  @behaviour Client.Spec
+
+  @impl true
+  def request(method, path, chain_id \\ "0", headers \\ [], body \\ "", opts \\ []),
```
Comment (juanhurtado10) [inline lib/chainweb/client.ex]: @santiagoboterokommit IMO, the client should not manage the `chain_id`, this is something that should be contained in the request path .... in other words, the API endpoints are responsible for handling it

File: lib/chainweb/util/network.ex [hunk 3]
```diff
@@ -0,0 +1,19 @@
+defmodule Kadena.Chainweb.Network do
```
Comment (juanhurtado10) [inline lib/chainweb/util/network.ex]: @santiagoboterokommit Are you sure this code will allow users to move between networks without changing environment variables?. Are you anticipating that?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/chainweb/util/network.ex [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
```
Comment (juanhurtado10) [inline lib/chainweb/util/network.ex]: @santiagoboterokommit please, remove this module since it is not related to the HTTP client impl.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: test/support/fixtures/chainweb/200.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+			"gasPrice": 0,
+			"sender": ""
+		},
+		"blockTime": 1666454535899148,
+		"prevBlockHash": "b3y6BM0IMBxPEubABoRWGatk6XY7FLoJIiezw-riz98",
+		"blockHeight": 2662676
+	},
+	"continuation": null,
+	"txId": null
+}
```
Comment (juanhurtado10) [inline test/support/fixtures/chainweb/200.json]: @santiagoboterokommit Please, fix the file indentation.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

**Review by juanhurtado10 — APPROVED** [review]

File: lib/chainweb/client/default.ex [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  @spec handle_response(response :: client_response()) :: parsed_response()
+  defp handle_response({:ok, status, _headers, body}) when status in 200..299 do
+    decoded_body = json_library().decode!(body, keys: :atoms)
+    {:ok, decoded_body}
+  end
+
+  defp handle_response({:ok, status, _headers, body}) when status in 400..599,
+    do: {:error, [chainweb: body]}
```
Comment (einerzg) [inline lib/chainweb/client/default.ex:34]: @santiagoboterokommit, It's not necessary to add this line here `decoded_body = json_library().decode!(body, keys: :atoms)`

**Review by einerzg — APPROVED** [review]

**Pull request merged**